src/backend/distributed/planner/shard_pruning.c

```diff
@@ -405,6 +523,150 @@ PruneShards(Oid relationId, Index rangeTableId, List *whereClauseList,
 }
 
 
+/*
+ * Check whether node is a valid constraint for pruning
+ */
+static bool
+IsValidConditionNode(Node *node, Var *partitionColumn)
+{
+	if (IsA(node, OpExpr))
+	{
+		OpExpr *opClause = (OpExpr *) node;
+		Var *varClause = NULL;
+		if (VarConstOpExprClause(opClause, partitionColumn, &varClause, NULL))
+		{
+			if (equal(varClause, partitionColumn))
+			{
+				return IsValidPartitionKeyRestriction(opClause);
+			}
+			else if (varClause->varattno == RESERVED_HASHED_COLUMN_ID)
+			{
+				return IsValidHashRestriction(opClause);
+			}
+		}
+
+		return false;
+	}
+	else if (IsA(node, ScalarArrayOpExpr))
+	{
+		ScalarArrayOpExpr *arrayOperatorExpression = (ScalarArrayOpExpr *) node;
+		if (SAORestrictions(arrayOperatorExpression, partitionColumn, NULL))
+		{
+			return true;
+		}
+
+		return false;
+	}
+	else
+	{
+		return false;
+	}
+}
+
+
+/*
+ * Build a logical tree of valid constraints and invalid constaints for pruning.
+ */
+static bool
+BuildPruningTree(Node *node, PruningTreeBuildContext *context)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, List))
+	{
+		return expression_tree_walker(node, BuildPruningTree, context);
+	}
+	else if (IsA(node, BoolExpr))
+	{
+		BoolExpr *boolExpr = (BoolExpr *) node;
+		bool isAnded = boolExpr->boolop == AND_EXPR;
+
+		if (boolExpr->boolop == NOT_EXPR)
+		{
+			return false;
```

Comment:
Should be able to descend & apply De Morgan's laws based on a bool in the context, eg

`WHERE NOT (id != 2 OR id != 3)` should give the same tree as `WHERE id = 2 AND id = 3`

Reply:
Postgres standard planner actually removes `NOT_EXPR`s. It behind the scenes already applies De Morgan's laws:
```
EXPLAIN SELECT * FROM orders_hash_partitioned
WHERE NOT (o_orderkey != 2 OR o_orderkey != 3);
                           QUERY PLAN
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)
   Task Count: 1
   Tasks Shown: All
   ->  Task
         Node: host=localhost port=xxxxx dbname=regression
         ->  Result  (cost=0.00..0.02 rows=1 width=346)
               One-Time Filter: false
               ->  Result  (cost=0.00..0.01 rows=1 width=346)
                     One-Time Filter: false
(9 rows)

EXPLAIN SELECT * FROM orders_hash_partitioned
WHERE NOT (o_orderkey = 2 OR o_orderkey = 3);
                                                      QUERY PLAN
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)
   Task Count: 4
   Tasks Shown: One of 4
   ->  Task
         Node: host=localhost port=xxxxx dbname=regression
         ->  Seq Scan on orders_hash_partitioned_630000 orders_hash_partitioned  (cost=0.00..13.15 rows=208 width=346)
               Filter: ((o_orderkey <> 2) AND (o_orderkey <> 3))
(7 rows)

EXPLAIN SELECT * FROM orders_hash_partitioned
WHERE NOT (o_orderkey != 2 AND o_orderkey != 3);
                                                     QUERY PLAN
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)
   Task Count: 2
   Tasks Shown: One of 2
   ->  Task
         Node: host=localhost port=xxxxx dbname=regression
         ->  Seq Scan on orders_hash_partitioned_630003 orders_hash_partitioned  (cost=0.00..13.15 rows=2 width=346)
               Filter: ((o_orderkey = 2) OR (o_orderkey = 3))
(7 rows)

EXPLAIN SELECT * FROM orders_hash_partitioned
WHERE NOT (o_orderkey = 2 AND o_orderkey = 3);
                                                      QUERY PLAN
---------------------------------------------------------------------
 Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0)
   Task Count: 4
   Tasks Shown: One of 4
   ->  Task
         Node: host=localhost port=xxxxx dbname=regression
         ->  Seq Scan on orders_hash_partitioned_630000 orders_hash_partitioned  (cost=0.00..13.15 rows=210 width=346)
               Filter: ((o_orderkey <> 2) OR (o_orderkey <> 3))
(7 rows)
```

So we could add a comment here that `NOT_EXPR` should never be encountered.

Reply:
Same thing happens for lt/gt/lteq/gteq:
´WHERE NOT (o_orderkey >= 2 OR o_orderkey <= 3)´ -> `Filter: ((o_orderkey < 2) AND (o_orderkey > 3))`

Also for `IN`:
´WHERE NOT o_orderkey IN (2, 3)´ -> `Filter: (o_orderkey <> ALL ('{2,3}'::integer[]))`

Reply:
Should we have better tests for NOT query cases also showing above behaviour? 🤔

Reply:
I did this now. I also marked NOT_EXPR as invalid since we should not encounter them